src/System.Diagnostics.DiagnosticSource/src/DiagnosticSourceUsersGuide.md

```diff
@@ -0,0 +1,377 @@
+to allow in process tools to get at very rich data.   Because the consumer is assumed to be
+within the same process, non-serializable types (e.g. HttpResponseMessage or HttpContext)
+can be passed, which gives the consumer a lot of potential data to work with.    
+
+As explained in [Bridge to EventSource](bridge-to-eventSource), there is a bridge that
+pipes information from DiagnosticSource's to an EventSource.   Thus EventSource
+consumers can get at all DiagnosticSource events.    While the data payloads from 
+DiagnosticSource can't in general be passed through to the EventSource (because they are 
+not serializable), there is a mechanism in the bridge that consumers can specify which fields
+(which are serializable) from the DiagnosticSource payload should be passed
+along to the EventSource.  
+
+What this means is that in general it is not necessary to instrument a code site multiple
+times.   By instrumenting with Diagnostic source, both clients that need the rich data
+(and thus use DiagnosticListener) as well as any consumers using EventListeners 
+(or OS facilities like ETW)) can get at the data.  
+
+----------------------------------------
+## Instrumenting with DiagnosticSource/DiagnosticListener
+
+Perhaps surprisingly, the heart of DiagnosticSource logging architecture is is not the
+DiagnositicSource class but rather the DiagnositicListener class which 'receives' the 
+events.    This is because the DiagnositicSource type is just an abstract base class
+that defines the methods needed to actually log events.    It is the  DiagnositicListener
+which holds all the actual implementation.
+
+Thus the first step in instrumenting code with DiagnosticSource is to create a 
+DiagnosticListener.   For example
+
+```C#
+    private static DiagnositicSource httpLogger = new DiagnositicListener("System.Net.Http");
+```
+Notice that httpLogger is typed as a DiagnositicSource.    This is because this code
+only cares about writing events and thus only cares about the  DiagnositicSource methods that
+the DiagnosticListener implements.   DiagnositicListeners are given names when they are created
+and this name should be the name of logical grouping of related events (typically the component).
+Later this name is used to find the Listener and subscribe to any of its events.  
+
+Once you have an instance of a DiagnositicSource, logging is very straightforward.  The
+interface consists of only two methods.  
+
+```C#
+	bool IsEnabled(string name)
+	void Write(string name, object value);
+```
+
+A typical call site will look like
+
+```C#
+	if (httpLogger.IsEnabled("RequestStart"))
+		httpLogger.Write("RequestStart", new { Url="http://clr", Request=aRequest });
+```
+
+Already some of the architectural elements are being exposed namely
+
+	1. Every event has a string name (e.g. Request Start), and exactly one object as a payload.
+	2. If you need to send more than one item, you can do so by creating a object with all information
```

Comment:
I had used C# instead of CSharp originally, but there seems to be some issue with using a # right before a URL (the []).  I did not care to figure it out so I just avoided the issue.